### PR TITLE
improve caching mechanism

### PR DESCRIPTION
**Pull request by StrikerRUS**
For now we can only reuse expression if it is the same piece of memory:

```
from m2cgen import ast
from m2cgen.interpreters import PythonInterpreter

expr1 = ast.BinNumExpr(ast.NumVal(1), ast.NumVal(1), ast.BinNumOpType.ADD, to_reuse=True)
expr2 = ast.BinNumExpr(ast.NumVal(1), ast.NumVal(1), ast.BinNumOpType.ADD, to_reuse=True)
expr = ast.BinNumExpr(expr1, expr2, ast.BinNumOpType.ADD)

interpreter = PythonInterpreter()

print(interpreter.interpret(expr))
```

```
def score(input):
    var0 = (1) + (1)
    var1 = (1) + (1)
    return (var0) + (var1)
```

For example, look how the amount of generated SVM code decreased with this improvement.

```
from sklearn.datasets import load_digits
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split

import m2cgen as m2c

X, y = load_digits(3, True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train = X_train[:15, :3]
y_train = y_train[:15]

est = SVC().fit(X_train, y_train)

print(m2c.export_to_python(est))
```

Before:
```
import math
def score(input):
    var0 = (0) - (0.026918168766948466)
    var1 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((1.0) - (input[2]), 2))))
    var2 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((0.0) - (input[2]), 2))))
    var3 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((1.0) - (input[2]), 2))))
    var4 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((0.0) - (input[2]), 2))))
    var5 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((3.0) - (input[2]), 2))))
    var6 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((3.0) - (input[2]), 2))))
    var7 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((11.0) - (input[2]), 2))))
    var8 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((2.0) - (input[2]), 2))))
    var9 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((3.0) - (input[2]), 2))))
    var10 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((0.0) - (input[2]), 2))))
    var11 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((10.0) - (input[2]), 2))))
    var12 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((9.0) - (input[2]), 2))))
    var13 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((7.0) - (input[2]), 2))))
    var14 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((3.0) - (input[1]), 2))) + (math.pow((14.0) - (input[2]), 2))))
    return [((((((((((0.05402739718556404) + ((var1) * (-1.0))) + ((var2) * (-1.0))) + ((var3) * (-1.0))) + ((var4) * (-0.8215458884546635))) + ((var5) * (-1.0))) + ((var6) * (1.0))) + ((var7) * (0.8215458884546635))) + ((var8) * (1.0))) + ((var9) * (1.0))) + ((var10) * (1.0)), (((((((((-0.15294075012207037) + ((var11) * (-0.17725803301371043))) + ((var12) * (-1.0))) + ((var13) * (-1.0))) + ((var14) * (-0.8227419669862895))) + ((var6) * (1.0))) + ((var7) * (1.0))) + ((var8) * (0.0))) + ((var9) * (1.0))) + ((var10) * (0.0)), (((((((((-0.20567373035050823) + ((var11) * (-0.0))) + ((var12) * (-0.17751524802252777))) + ((var13) * (-1.0))) + ((var14) * (-0.5487240974910604))) + ((var1) * (0.0))) + ((var2) * (0.0))) + ((var3) * (0.6054875767916933))) + ((var4) * (0.12075176872189494))) + ((var5) * (1.0))]
```

After:
```
import math
def score(input):
    var0 = (0) - (0.026918168766948466)
    var1 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((1.0) - (input[2]), 2))))
    var2 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((0.0) - (input[2]), 2))))
    var3 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((3.0) - (input[2]), 2))))
    var4 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((11.0) - (input[2]), 2))))
    var5 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((2.0) - (input[2]), 2))))
    var6 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((10.0) - (input[2]), 2))))
    var7 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((9.0) - (input[2]), 2))))
    var8 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((0.0) - (input[1]), 2))) + (math.pow((7.0) - (input[2]), 2))))
    var9 = math.exp((var0) * (((math.pow((0.0) - (input[0]), 2)) + (math.pow((3.0) - (input[1]), 2))) + (math.pow((14.0) - (input[2]), 2))))
    return [((((((((((0.05402739718556404) + ((var1) * (-1.0))) + ((var2) * (-1.0))) + ((var1) * (-1.0))) + ((var2) * (-0.8215458884546635))) + ((var3) * (-1.0))) + ((var3) * (1.0))) + ((var4) * (0.8215458884546635))) + ((var5) * (1.0))) + ((var3) * (1.0))) + ((var2) * (1.0)), (((((((((-0.15294075012207037) + ((var6) * (-0.17725803301371043))) + ((var7) * (-1.0))) + ((var8) * (-1.0))) + ((var9) * (-0.8227419669862895))) + ((var3) * (1.0))) + ((var4) * (1.0))) + ((var5) * (0.0))) + ((var3) * (1.0))) + ((var2) * (0.0)), (((((((((-0.20567373035050823) + ((var6) * (-0.0))) + ((var7) * (-0.17751524802252777))) + ((var8) * (-1.0))) + ((var9) * (-0.5487240974910604))) + ((var1) * (0.0))) + ((var2) * (0.0))) + ((var1) * (0.6054875767916933))) + ((var2) * (0.12075176872189494))) + ((var3) * (1.0))]
```

### CODE REVIEW

File: m2cgen/interpreters/interpreter.py [hunk 1]
```diff
@@ -29,6 +29,9 @@ def _do_interpret(self, expr, to_reuse=None, **kwargs):
         if result is not None:
             return result
 
+        if expr in self._cached_expr_results:
+            return self._cached_expr_results[expr].var_name
```
Comment (StrikerRUS) [inline m2cgen/interpreters/interpreter.py:33]: Search in cache for expressions that are now not `to_reuse` but might be cached earlier.

**Comment by izeigerman** [comment]
That's a fantastic improvement and I was thinking about it myself. My only concern is that this approach will only work for referentially transparent expressions ([link](https://en.wikipedia.org/wiki/Referential_transparency)). The question I had in mind was whether it's always going to be the case or can our expressions have side effects in future. What do you think? Am I missing a point here?

**Comment by StrikerRUS** [comment]
> The question I had in mind was whether it's always going to be the case or can our expressions have side effects in future.

I think that present benefits (by caching and reducing code size we are contributing to #88) outperform possible future inconveniences. I remember we had similar discussion https://github.com/BayesWitnesses/m2cgen/pull/199#discussion_r421635242 which led to introducing `IdExpr`. So I believe we will be able to invent some workarounds again if needed. For now, I don't see any fundamental changes in the codebase that may require making expressions impure.

File: m2cgen/interpreters/mixins.py [hunk 2]
```diff
@@ -36,10 +36,10 @@ def _pre_interpret_hook(self, expr, bin_depth=0, **kwargs):
 
     # Default implementation. Simply adds new variable.
     def bin_depth_threshold_hook(self, expr, **kwargs):
-        var_name = self._cg.add_var_declaration(expr.output_size)
+        if expr in self._cached_expr_results:
+            return self._cached_expr_results[expr].var_name
         result = self._do_interpret(expr, **kwargs)
-        self._cg.add_var_assignment(var_name, result, expr.output_size)
-        return var_name
+        return self._cache_reused_expr(expr, result)
```
Comment (izeigerman) [inline m2cgen/interpreters/mixins.py:42]: I don't think we should mix 2 cases together. We have 2 scenarios to consider here:
1. An expression that is explicitly marked as reusable (`to_reuse` is true). This is what we call "caching" in the code.
2. An AST requires splitting due to limitation of the target platform / langauge.

I can easily see how a specific interpreter may choose to opt into one of these mechanisms but not the other. I can also see how 2 of these may require different implementations (eg. new variable vs new function). Additionally by default the caching is disabled (https://github.com/BayesWitnesses/m2cgen/blob/892d350d114ac81696522c41c1498ebc2c8f1b28/m2cgen/interpreters/interpreter.py#L49-L51), which may be quite unexpected if one chooses to rely on `BinExpressionDepthTrackingMixin`.

File: m2cgen/ast.py [hunk 3]
```diff
@@ -63,6 +81,12 @@ def __str__(self):
         args = ",".join([str(self.expr), "to_reuse=" + str(self.to_reuse)])
         return "AbsExpr(" + args + ")"
 
+    def __eq__(self, other):
+        return type(other) is AbsExpr and self.expr == other.expr
```
Comment (izeigerman) [inline m2cgen/ast.py:85]: It bothers me semantically that we don't check the `to_reuse` flag for equality too.

File: m2cgen/ast.py [hunk 4]
```diff
@@ -164,6 +223,14 @@ def __str__(self):
         args = ",".join([str(e) for e in self.exprs])
         return "VectorVal([" + args + "])"
 
+    def __eq__(self, other):
+        return (type(other) is VectorVal and
+                self.output_size == other.output_size and
+                all(i == j for i, j in zip(self.exprs, other.exprs)))
+
+    def __hash__(self):
+        return hash(tuple(e for e in self.exprs))
```
Comment (izeigerman) [inline m2cgen/ast.py:232]: I might be missing some important detail here but why can't we just pass `self.exprs` into `tuple` directly? I.e.: `tuple(self.exprs)`

File: m2cgen/ast.py [hunk 5]
```diff
@@ -144,6 +194,15 @@ def __str__(self):
         ])
         return "BinNumExpr(" + args + ")"
 
+    def __eq__(self, other):
+        return (type(other) is BinNumExpr and
```
Comment (izeigerman) [inline m2cgen/ast.py:198]: This block of code looks pretty repetitive. Can we generalize it? Eg.:
```
def _eq_bin_exprs(expr_self, expr_other, expected_type):
        return (type(expr_self) is expected_type and
                expr_self.left == expr_other.left and
                expr_self.right == expr_other.right and
                expr_self.op == expr_other.op)

def __eq__(self, other):
    return _eq_bin_exprs(self, other, BinNumExpr)
```

File: m2cgen/interpreters/mixins.py (hunk 2 above)
Comment (StrikerRUS) [reply m2cgen/interpreters/mixins.py]: Absolutely agree with you! Refactored.

File: m2cgen/ast.py (hunk 4 above)
Comment (StrikerRUS) [reply m2cgen/ast.py]: Fixed, thanks!

File: m2cgen/ast.py (hunk 5 above)
Comment (StrikerRUS) [reply m2cgen/ast.py]: Refactored.

File: m2cgen/ast.py (hunk 3 above)
Comment (StrikerRUS) [reply m2cgen/ast.py:85]: I believe it is OK to not carry about unimportant attributes of objects during comparison. Adding `to_reuse` flag here will eliminate practically the whole point of this PR.

File: m2cgen/ast.py [hunk 6]
```diff
@@ -233,6 +309,15 @@ def __str__(self):
         args = ",".join([str(self.left), str(self.right), self.op.name])
         return "CompExpr(" + args + ")"
 
+    def __eq__(self, other):
+        return (type(other) is CompExpr and
```
Comment (izeigerman) [inline m2cgen/ast.py:313]: Can we use `_eq_bin_exprs` here too?

Comment (StrikerRUS) [reply m2cgen/ast.py:313]: Technically yes. But `CompExpr` doesn't inherit `BinExpr`, so I'm a little bit confused about semantics. However, I don't have strong opinion here.

Comment (izeigerman) [reply m2cgen/ast.py:313]: Well technically it's still a binary expression, just not a numeric one.

Comment (StrikerRUS) [reply m2cgen/ast.py]: Done in f7b7309a31689ecdefd3789dbf6ee065f7a6336d.

**Pull request merged**